Mosek Quadratic Programming

alexdunyak opened the pull request:
The code is very similar to the linear programming code. I renamed MosekLP to MosekInterface, as there would be no need for multiple classes on closer inspection. The bulk of the new code is in the Solve function of MosekInterface, where it just uses a conditional on the the "problemtype" solver options to distinguish between linear and quadratic programming, and mostly it marshalls data to the newly updated class constructor . The included unit tests work in release mode on my computer, so hopefully they will pass CI promptly.

If you have any comments, don't hesitate to share them.

Thanks,
Alex Dunyak

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2894)
<!-- Reviewable:end -->

alexdunyak commented:
Uh does anyone know why that happened?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNIaoFpqK8ksR1sLB4E:1716079787)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
ZOMG Wow.  Good job jenkins.

@drake-jenkins-bot ok to test

jwnimmer-tri commented:
@jamiesnape any way to prevent the the jenkins spam above?

jamiesnape commented:
Looks like the idea of having on-demand Jenkins jobs triggered by comments is not going to work.

jamiesnape commented:
I may be able to patch the plugin. Can you live with the noise on this thread while I work on it?

alexdunyak commented:
It's fine with me.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNIhYPBv-J5CplSRWa8:1797347326)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@jamiesnape Yeah, noise here or otherwise is fine, just wanted to make sure it was on your radar.

ggould-tri commented:
Apologies to @alexdunyak that this sat in the queue for so long without anyone noticing (except for @drake-jenkins-bot, that old lush).

+@ggould-tri feature review
+@sherm1 platform review

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNYFDiPhOVG8ihvspJQ:-1640050631)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Thought I fixed @drake-jenkins-bot. I guess not.

ggould-tri commented:
@jamiesnape looks like @drake-jenkins-bot has gone crazy on us again.  Any idea why?

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNYGjCWt99aszNYT1aA:354669122)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@drake-jenkins-bot ok to test
Goddamnit.

jamiesnape commented:
@drake-jenkins-bot add to whitelist

jamiesnape commented:
Trying temporarily closing the PR to silence @drake-jenkins-bot.

jamiesnape commented:
I think he is in an infinite loop.

jwnimmer-tri commented:
For the record -- the bot user `qenxr-wraxvaf-obg | rot13` (name obscured to avoid beetlejuicing him back into existence) posted ~250 comments to this PR asking to verify the patch.  I deleted all but a few of them.

jamiesnape commented:
The problem was that every comment was causing him to ask for permission. Will reopen this once patch attempt 2 is deployed.

jwnimmer-tri commented:
@jamiesnape Can we reopen this yet?

jamiesnape commented:
The infinite loop has been sorted, but still working on the previous issue. When you re-open, you are going to get ~30 comments, after that none unless a new commit is added.

jamiesnape commented:
Or maybe just one. Cool.

alexdunyak commented:
Since it passed CI, can  @ggould-tri and @sherm1 take a look?

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNh6A6dCvOEa3J38e8j:-977212353)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
I'll get on this this afternoon.

On Wed, Jul 27, 2016 at 12:08 PM, alexdunyak <notifications@github.com>
wrote:

> Since it passed CI, can @ggould-tri <https://github.com/ggould-tri> and
> @sherm1 <https://github.com/sherm1> take a look?
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNh6A6dCvOEa3J38e8j:-977212353>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/2894#issuecomment-235635289>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQx__XBhQ16qszVqbNKfnLnWE92_HdD5ks5qZ4KAgaJpZM4JS85R>
> .
>



-- 
Grant Gould
Toyota Research Institute

sherm1 commented:
I haven't looked at this so I'm passing off platform review to today's platform reviewer, +@sammy-tri, -@sherm1.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNhQ0x3l1nqnbfGb1VW:-1887372771)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 1 of 2 files at r2.
Review status: 1 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/solvers/MosekInterface.h, line 34 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhUZ9ztO3xP-f0fxru:-KNhUZ9ztO3xP-f0fxrv:-2094444796) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L34)):*
> ```objc
>  *  It is created by a MosekSolver object.
>  */
> class DRAKEOPTIMIZATION_EXPORT MosekInterface {
> ```

This class name conflicts with the style guide, which says only pure interfaces should end with the "Interface" suffix: https://google.github.io/styleguide/cppguide.html#Interfaces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894)*
<!-- Sent from Reviewable.io -->

alexdunyak commented:
*[drake/solvers/MosekInterface.h, line 34 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhUZ9ztO3xP-f0fxru:-KNhZvBnGbdtZ5rq79hZ:556578746) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L34)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This class name conflicts with the style guide, which says only pure interfaces should end with the "Interface" suffix: https://google.github.io/styleguide/cppguide.html#Interfaces
</details>
Would it be better to rename the class to MosekClass (or some other, better name if you have one) or to fold functionality into MosekSolver?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 5 of 8 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/solvers/MosekInterface.cpp, line 5 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWnFtGbKdqXxhS2V0:-KNhWnFtGbKdqXxhS2V1:950499839) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L5)):*
> ```cpp
> #include "drake/solvers/MosekInterface.h"
> 
> #include <memory>
> ```

Includes should be ordered alphabetically within sections: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

---

*[drake/solvers/MosekInterface.cpp, line 7 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWW7tF-H9o99yZcjy:-KNhWW7tF-H9o99yZcjz:1189719377) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L7)):*
> ```cpp
> #include <memory>
> #include <vector>
> #include <iostream>
> ```

Appears to be unused.

---

*[drake/solvers/MosekInterface.cpp, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWeLNAsFnSatzKscI:-KNhWeLNAsFnSatzKscJ:1189719377) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L9)):*
> ```cpp
> #include <iostream>
> #include <string>
> #include <map>
> ```

Appears to be unused.

---

*[drake/solvers/MosekInterface.cpp, line 11 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhXamnyzwAgSSIvAI7:-KNhXamo0Wv6AZde5fga:950499839) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L11)):*
> ```cpp
> #include <map>
> 
> #include <Eigen/Sparse>
> ```

Includes should be ordered alphabetically within sections: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

---

*[drake/solvers/MosekInterface.cpp, line 69 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhYdMQXnL5MsOGJ3I2:-KNhYdMRZdEb_ctwN3fo:1231503722) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L69)):*
> ```cpp
> }
> 
> void MosekInterface::AddQuadraticObjective(Eigen::MatrixXd quad_objective) {
> ```

We should pass potentially large data types like matrices by reference rather than by value.

---

*[drake/solvers/MosekInterface.cpp, line 87 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhYrPg4HMwDKJGai1q:-KNhYrPhL0NUaDLDvnlR:1231503722) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L87)):*
> ```cpp
> }
> 
> void MosekInterface::AddQuadraticConstraintMatrix(Eigen::MatrixXd quad_cons) {
> ```

We should pass potentially large data types like matrices by reference rather than by value.

---

*[drake/solvers/MosekInterface.cpp, line 157 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZ7XT2LGD1-dgRyGz:-KNhZ7XT2LGD1-dgRyH-:-675289819) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L157)):*
> ```cpp
>     const std::vector<double>& upper_bounds,
>     const std::vector<double>& lower_bounds) {
>   int j = 0;
> ```

Why not just declare j inside the for loop initializer?

---

*[drake/solvers/MosekInterface.cpp, line 197 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZKIM95M97gPUzZ6n:-KNhZKIM95M97gPUzZ6o:-185376806) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L197)):*
> ```cpp
>   // something at least.
>   // assume that the problem type is linear currently.
>   // TODO(adunyak): Add support for quadratic objective and constraints
> ```

This seems to be done now.

---

*[drake/solvers/MosekInterface.cpp, line 233 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZQj6hq7NyV4J2Dne:-KNhZQj6hq7NyV4J2Dnf:533432518) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L233)):*
> ```cpp
>   }
>   mosek_variable_bounds = FindMosekBounds(upper_variable_bounds,
>                                          lower_variable_bounds);
> ```

This looks like it should be indented another space?

---

*[drake/solvers/MosekInterface.cpp, line 255 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgPBYtkRDITWnLgzo:-KNhgPBYtkRDITWnLgzp:1520896949) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L255)):*
> ```cpp
>     upper_constraint_bounds.resize(totalconnum);
>     lower_constraint_bounds.resize(totalconnum);
>     // mosek_constraint_bounds.resize(totalconnum);
> ```

Leftover (or incorrectly) commented out code

---

*[drake/solvers/MosekInterface.cpp, line 282 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgdg2RDYQZQnvTK7-:-KNhgdg3W6zkHaQCf3uM:-1942060819) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L282)):*
> ```cpp
>                                             lower_constraint_bounds);
>     // find the linear objective here
>     LinearConstraint *obj = dynamic_cast<LinearConstraint *>(
> ```

I'm not thrilled about the dynamic_cast, but I think we should at least check that generic_costs() is of size 1 before trying to use the output of front().

---

*[drake/solvers/MosekInterface.cpp, line 313 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhf0rI9zZWYhjV1e_W:-KNhf0rI9zZWYhjV1e_X:1889793916) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L313)):*
> ```cpp
>     // TODO(alexdunyak): Allow multiple quadratic constraints
>     DRAKE_ASSERT(prog.generic_constraints().size() == 1);
>     QuadraticConstraint *quad_con_ptr = dynamic_cast<QuadraticConstraint *>(
> ```

As the comment above says, there's no getter for quadratic constraints yet.  I believe the code here is correct and sufficient to test that Mosek integration is working.

It will not be sufficient for the next stage of Mosek integration, when we want to be able to call MosekSolver through OptimizationProblem/MathematicalProgram.

I'd suggest a TODO comment which looks like the following:

"Once we've added OptimizationProblem::AddQuadraticConstraint and started to store quadratic constraints independently of generic constraints, this cast should go away and we should get the quadratic constraints from the OptimizationProblem without a cast being necessary.  Make sure to assert that other types of constraints are not also present."

---

*[drake/solvers/MosekInterface.cpp, line 316 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgBsIOrd_k4vbNZ9K:-KNhgBsIOrd_k4vbNZ9L:-434511984) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L316)):*
> ```cpp
>         prog.generic_constraints().front().constraint().get());
>     DRAKE_ASSERT(quad_con_ptr != nullptr);
>     linear_cons = (*quad_con_ptr).b().transpose();
> ```

This syntax seems strange compared to "quad_con_ptr->b().transpose()"

---

*[drake/solvers/MosekInterface.cpp, line 340 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgs_9fQLE3rsy7HsA:-KNhgs_9fQLE3rsy7HsB:511867383) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L340)):*
> ```cpp
> 
>     // Assume only one quadratic objective over the whole problem.
>     DRAKE_ASSERT(prog.quadratic_costs().size() == 1);
> ```

Should we also assert that generic costs is empty?

---

*[drake/solvers/MosekInterface.cpp, line 372 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgxoUZsliSokcx-Sp:-KNhgxoUZsliSokcx-Sq:1901129709) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L372)):*
> ```cpp
> 
> SolutionResult MosekInterface::OptimizeTask(const std::string& maxormin,
>                                      const std::string& ptype) {
> ```

Unusual indentation

---

*[drake/solvers/MosekInterface.h, line 30 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhVnHoXRoSFGsktpTE:-KNhVnHpm_fxCIF2mFal:-1770084078) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L30)):*
> ```objc
>  *  Specifically, the program options:
>  *  -- "maxormin" -- must be set to "max" or "min"
>  *  -- "problemtype" -- must be set to "linear"
> ```

This comment appears to be out of date now that quadratic is supported.

---

*[drake/solvers/MosekInterface.h, line 34 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhUZ9ztO3xP-f0fxru:-KNhiJEn_y8uElfJF9Tp:-1686255037) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L34)):*
<details><summary><i>Previously, alexdunyak wrote…</i></summary>
> Would it be better to rename the class to MosekClass (or some other, better name if you have one) or to fold functionality into MosekSolver?
</details>
MosekWrapper might be OK?  You could merge it into MosekSolver, but I'm not sure it's worth the architectural change at this point.  I'd take other suggestions for names which aren't Interface...

---

*[drake/solvers/MosekInterface.h, line 54 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhW7VS-7f1efqYFtlC:-KNhW7VS-7f1efqYFtlD:-420235100) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L54)):*
> ```objc
>     std::vector<double> upper_variable_bounds,
>     std::vector<double> lower_variable_bounds,
>     double constant_eqn_term = 0,
> ```

If we only create these objects from a single point in the Solve() function, could we get rid of the default arguments by specifying the values either in Solve() or in the constructor?

---

*[drake/solvers/test/testMosek.cpp, line 74 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhCDjMvOTcxcRjgcj:-KNhhCDkiGA5624e-9gM:271860960) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L74)):*
> ```cpp
>   OptimizationProblem prog2;
>   auto x = prog2.AddContinuousVariables(3);
>   // Build the objective
> ```

Comments should have proper punctuation and grammar (there are a few more of these in this function): https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar

---

*[drake/solvers/test/testMosek.cpp, line 82 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhcu12pS8bhf0xthJ:-KNhhcu12pS8bhf0xthK:-325483836) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L82)):*
> ```cpp
>   c << 0, -1, 0;
>   QuadraticConstraint obj(Q, c, 0, 0);
>   prog2.AddQuadraticCost(std::make_shared<QuadraticConstraint>(obj));
> ```

Creating an object on the stack, bouncing it through the copy constructor implicitly in make_shared (I think that's what's happening here?) and then passing that shared_ptr into AddQuadraticCost seems much more complicated than just calling the version of AddQuadraticCost which takes the Q and b parameters directly and makes the shared_ptr for you.

---

*[drake/solvers/test/testMosek.cpp, line 89 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhwxIkY_wHsUHb27D:-KNhhwxJxZBLoHl7TFxt:474739653) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L89)):*
> ```cpp
>       std::numeric_limits<double>::infinity());
>   std::shared_ptr<QuadraticConstraint> ptrtocon =
>       std::make_shared<QuadraticConstraint>(con);
> ```

Using the copy constructor (I think that's what's happening?) to make a shared ptr from a stack variable seems complicated.  Why not pass the parameters to con's constructor into the make_shared call and avoid building a stack allocated version?

---

*[drake/solvers/test/testMosek.cpp, line 122 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhiA908Ny8r3xJa6Lq:-KNhiA91GpHCs450o780:-70215498) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L122)):*
> ```cpp
>   c << 0, -1, 0;
>   QuadraticConstraint obj(Q, c, 0, 0);
>   prog2.AddQuadraticCost(std::make_shared<QuadraticConstraint>(obj));
> ```

See above about copying stack allocated items to make shared pointers.

---

*[drake/solvers/test/testMosek.cpp, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhiEYAXFkqecrahcAQ:-KNhiEYAXFkqecrahcAR:-70215498) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L133)):*
> ```cpp
>       std::numeric_limits<double>::infinity());
>   std::shared_ptr<QuadraticConstraint> ptrtocon =
>       std::make_shared<QuadraticConstraint>(con);
> ```

See above about copying stack allocated items to make shared pointers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894)*
<!-- Sent from Reviewable.io -->

alexdunyak commented:
Review status: 2 of 7 files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/solvers/MosekWrapper.cpp, line 5 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWnFtGbKdqXxhS2V0:-KNhwb5KRue83zXIhi6I:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekWrapper.cpp#L5)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Includes should be ordered alphabetically within sections: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
</details>
Done.

---

*[drake/solvers/MosekWrapper.h, line 34 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhUZ9ztO3xP-f0fxru:-KNhwckqemPK4kvWQwIp:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekWrapper.h#L34)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> MosekWrapper might be OK?  You could merge it into MosekSolver, but I'm not sure it's worth the architectural change at this point.  I'd take other suggestions for names which aren't Interface...
</details>
Done.

---

*[drake/solvers/test/testMosek.cpp, line 74 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhCDjMvOTcxcRjgcj:-KNhwdujfNxpHGmR0A4p:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L74)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Comments should have proper punctuation and grammar (there are a few more of these in this function): https://google.github.io/styleguide/cppguide.html#Punctuation,_Spelling_and_Grammar
</details>
Done.

---

*[drake/solvers/test/testMosek.cpp, line 82 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhcu12pS8bhf0xthJ:-KNho2D5NumrxpJcxXFM:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L82)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Creating an object on the stack, bouncing it through the copy constructor implicitly in make_shared (I think that's what's happening here?) and then passing that shared_ptr into AddQuadraticCost seems much more complicated than just calling the version of AddQuadraticCost which takes the Q and b parameters directly and makes the shared_ptr for you.
</details>
Done.

---

*[drake/solvers/test/testMosek.cpp, line 89 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhhwxIkY_wHsUHb27D:-KNho2tPNi0IsLWDyvNT:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L89)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Using the copy constructor (I think that's what's happening?) to make a shared ptr from a stack variable seems complicated.  Why not pass the parameters to con's constructor into the make_shared call and avoid building a stack allocated version?
</details>
Done.

---

*[drake/solvers/test/testMosek.cpp, line 122 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhiA908Ny8r3xJa6Lq:-KNho4619cjL87gr5rTK:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L122)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> See above about copying stack allocated items to make shared pointers.
</details>
Done.

---

*[drake/solvers/test/testMosek.cpp, line 133 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhiEYAXFkqecrahcAQ:-KNhp_bNURt704cbtjrZ:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/test/testMosek.cpp#L133)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> See above about copying stack allocated items to make shared pointers.
</details>
Done.

---

*[drake/solvers/MosekInterface.cpp, line 7 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWW7tF-H9o99yZcjy:-KNhkCEeg0qghaczq1fz:-1974277881) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L7)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Appears to be unused.
</details>
Done. Removed.

---

*[drake/solvers/MosekInterface.cpp, line 9 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhWeLNAsFnSatzKscI:-KNhkE7JNsHHBpjIGnrw:-1974277881) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L9)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Appears to be unused.
</details>
Done. Removed.

---

*[drake/solvers/MosekInterface.cpp, line 11 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhXamnyzwAgSSIvAI7:-KNhkFlWn8mYz55p3aKX:-1862153331) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L11)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Includes should be ordered alphabetically within sections: https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes
</details>
Done. Reordered.

---

*[drake/solvers/MosekInterface.cpp, line 69 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhYdMQXnL5MsOGJ3I2:-KNhkS-rVdZo64c9x3kT:-281349707) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L69)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> We should pass potentially large data types like matrices by reference rather than by value.
</details>
Done. Passed by reference.

---

*[drake/solvers/MosekInterface.cpp, line 87 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhYrPg4HMwDKJGai1q:-KNhkV5bdvSqNXTL5nF7:857090948) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L87)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> We should pass potentially large data types like matrices by reference rather than by value.
</details>
Done. Passes by reference.

---

*[drake/solvers/MosekInterface.cpp, line 157 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZ7XT2LGD1-dgRyGz:-KNhkgIOwCzwwdm6VrNu:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L157)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Why not just declare j inside the for loop initializer?
</details>
Done.

---

*[drake/solvers/MosekInterface.cpp, line 197 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZKIM95M97gPUzZ6n:-KNhktMZFSkRrEs8aSJ6:-1598075736) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L197)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This seems to be done now.
</details>
Done. Comment removed.

---

*[drake/solvers/MosekInterface.cpp, line 233 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhZQj6hq7NyV4J2Dne:-KNhkwJvAdcrpflvuQPm:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L233)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This looks like it should be indented another space?
</details>
Done.

---

*[drake/solvers/MosekInterface.cpp, line 255 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgPBYtkRDITWnLgzo:-KNhkx13ZfzQBaYI0PX9:1844460769) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L255)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Leftover (or incorrectly) commented out code
</details>
Done. Line removed.

---

*[drake/solvers/MosekInterface.cpp, line 282 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgdg2RDYQZQnvTK7-:-KNhlMh1L-0XKyUH7buJ:2090414154) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L282)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> I'm not thrilled about the dynamic_cast, but I think we should at least check that generic_costs() is of size 1 before trying to use the output of front().
</details>
Done. DRAKE_ASSERT added.

The reason that the dynamic cast is necessary is because there is currently no way to get a linear objective from an OptimizationProblem object without either denoting it as a specific linear constraint.  

---

*[drake/solvers/MosekInterface.cpp, line 313 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhf0rI9zZWYhjV1e_W:-KNhmyKGek8Lz4kHGTTR:-898737807) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L313)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> As the comment above says, there's no getter for quadratic constraints yet.  I believe the code here is correct and sufficient to test that Mosek integration is working.
> 
> It will not be sufficient for the next stage of Mosek integration, when we want to be able to call MosekSolver through OptimizationProblem/MathematicalProgram.
> 
> I'd suggest a TODO comment which looks like the following:
> 
> "Once we've added OptimizationProblem::AddQuadraticConstraint and started to store quadratic constraints independently of generic constraints, this cast should go away and we should get the quadratic constraints from the OptimizationProblem without a cast being necessary.  Make sure to assert that other types of constraints are not also present."
> 
</details>
Done. TODO added. 

---

*[drake/solvers/MosekInterface.cpp, line 316 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgBsIOrd_k4vbNZ9K:-KNhn3RYh2VMwQMzkcSk:-1509572546) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L316)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This syntax seems strange compared to "quad_con_ptr->b().transpose()"
</details>
Done. Syntax changed.

---

*[drake/solvers/MosekInterface.cpp, line 340 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgs_9fQLE3rsy7HsA:-KNhnQiTmzy0omEzfmiw:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L340)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Should we also assert that generic costs is empty?
</details>
Done.

---

*[drake/solvers/MosekInterface.cpp, line 372 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhgxoUZsliSokcx-Sp:-KNhnYajjjztmUmLv8mb:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.cpp#L372)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> Unusual indentation
</details>
Done.

---

*[drake/solvers/MosekInterface.h, line 30 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhVnHoXRoSFGsktpTE:-KNhni5P-70K_WXGJ50P:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L30)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> This comment appears to be out of date now that quadratic is supported.
</details>
Done.

---

*[drake/solvers/MosekInterface.h, line 54 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNhW7VS-7f1efqYFtlC:-KNhnoy8sBxfk-w7ajqw:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/6a4e2e9ec652655be69f13d95f8a35f6e26aa06d/drake/solvers/MosekInterface.h#L54)):*
<details><summary><i>Previously, sammy-tri wrote…</i></summary>
> If we only create these objects from a single point in the Solve() function, could we get rid of the default arguments by specifying the values either in Solve() or in the constructor?
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending Jenkins and @ggould-tri's review.

---

Reviewed 1 of 10 files at r1, 5 of 7 files at r4.
Review status: 6 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNi-ef_p7h6olVjuBEr:-2057569069)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Reviewed 2 of 7 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
I've noted a few syntactic issues (the ones flagged 'No defect' are purely discretionary).  I am applying my <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> so that you can merge this as soon as the rest are fixed.

Thanks again for your work!

---

Reviewed 1 of 10 files at r1, 1 of 2 files at r2, 3 of 7 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[drake/solvers/MosekWrapper.cpp, line 37 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi57ZCZmEzXRozxt3S:-KNi57ZCZmEzXRozxt3T:-232934542) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L37)):*
> ```cpp
>   task_ = NULL;
>   solutions_.clear();
>   r_ = MSK_makeenv(&env_, NULL);
> ```

FYI -- the assignments above could be done more idiomatically with an initializer list.  However this is not required by our code conventions.

No defect.

---

*[drake/solvers/MosekWrapper.cpp, line 117 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi5ZpLDaQdzmhDtFdW:-KNi5ZpLDaQdzmhDtFdX:2043176534) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L117)):*
> ```cpp
>   int j = 0;  // iterator
>   // Define sparse matrix representation to be the same size as the desired
>   // constraints
> ```

FYI -- complete sentence requires `.` at the end.

---

*[drake/solvers/MosekWrapper.cpp, line 133 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi5eAYNcATqfJhkASh:-KNi5eAYNcATqfJhkASi:1470129247) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L133)):*
> ```cpp
> 
>   // following code adapted from http://docs.mosek.com/7.1/capi/Linear_optimization.html
>   // check if still working in valid environment
> ```

FYI -- This sentence is not grammatical; please rewrite as a complete sentence, capitalize, and punctuate.

---

*[drake/solvers/MosekWrapper.cpp, line 150 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi5qmH8NzsC_jsTzN7:-KNi5qmH8NzsC_jsTzN8:1348959020) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L150)):*
> ```cpp
>   for (i = 0; i < numcon_ && r_ == MSK_RES_OK; i++) {
>     r_ = MSK_putconbound(task_, i, mosek_bounds[i],
>                         lower_bounds[i], upper_bounds[i]);
> ```

FYI -- this indentation looks off-by-one.

---

*[drake/solvers/MosekWrapper.cpp, line 162 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi5vMVpDaF2qjM5amb:-KNi5vMVpDaF2qjM5amc:243456788) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L162)):*
> ```cpp
>   for (int j = 0; j < numvar_ && r_ == MSK_RES_OK; j++) {
>     r_ = MSK_putvarbound(task_, j, mosek_bounds[j], lower_bounds[j],
>         upper_bounds[j]);
> ```

FYI, I think this indentation may be wrong.

---

*[drake/solvers/MosekWrapper.cpp, line 196 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi62HX5CcP5eYm0tHj:-KNi62HYUOFekasam4jK:-1736409997) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L196)):*
> ```cpp
>   // construct an object that calls all the previous work so I can salvage
>   // something at least.
>   // assume that the problem type is linear currently.
> ```

FYI -- Please capitalize the two sentences above.

---

*[drake/solvers/MosekWrapper.cpp, line 316 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi6IQznDfWYaT7C_4M:-KNi6IQznDfWYaT7C_4N:-1619659403) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.cpp#L316)):*
> ```cpp
>     // use generic constraints and dynamically cast them to quadratic
>     // constraints.
>     // For now assume a single quadratic constraint
> ```

FYI -- Complete sentence requires terminal `.`.

---

*[drake/solvers/MosekWrapper.h, line 56 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi3eH5ZDppZ586psDR:-KNi3eH65bxY1oCKWxlW:1096637040) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/MosekWrapper.h#L56)):*
> ```objc
>     double constant_eqn_term,
>     Eigen::MatrixXd quad_objective,
>     Eigen::MatrixXd quad_cons);
> ```

FYI -- I was surprised to find that our code conventions do not require all of these to be `const&`.  I don't think the extra copy here is particularly expensive, but in the future you should be aware that `const SomeNontrivialStructure&` is computationally faster and more idiomatic.

No defect.

---

*[drake/solvers/Optimization.h, line 757 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi2FWInpgHh-eFSBZg:-KNi2FWInpgHh-eFSBZh:-1640824198) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/Optimization.h#L757)):*
> ```objc
>    *  either "max" or "min"
>    * -- "problemtype" -- currently only accepts "linear" or "quadratic". Use the
>    *  appropriate one.
> ```

FYI -- if you formatted this as markdown style, doxygen would be prettier.  However  this is not required.

No defect.

---

*[drake/solvers/Optimization.h, line 758 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-KNi2DPlC4jJvXWJy1s0:-KNi2DPlC4jJvXWJy1s1:-11725945) ([raw file](https://github.com/robotlocomotion/drake/blob/5dfef86171fa1ba468bbbf7e6ac950d727a6d7ba/drake/solvers/Optimization.h#L758)):*
> ```objc
>    * -- "problemtype" -- currently only accepts "linear" or "quadratic". Use the
>    *  appropriate one.
>    * TODO: Calling OptimizationProblem::Solve will not invoke mosek at this time.
> ```

FYI -- Long line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNi6W0052Njzjxiqrnm:1921349385)*
<!-- Sent from Reviewable.io -->

alexdunyak commented:
Hey, all the checks and such have passed, but the merge option is till grayed out for me.  Any idea why that might be?

Thanks for all the help!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNmPvLIV7pJZ-jiPbtw:-1173146822)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@alexdunyak You do not have permissions to merge.

ggould-tri commented:
I'll push the button  for you :-)

---

Review status: 4 of 7 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNmQr4D9fbD7Uk1lIIG:1736428726)*
<!-- Sent from Reviewable.io -->

alexdunyak commented:
Ok, thanks!

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2894#-:-KNmRKr6hF_EQCOMj0BJ:-309556765)*
<!-- Sent from Reviewable.io -->
